src/core/global.c

```diff
@@ -570,27 +563,46 @@ int nn_socket (int domain, int protocol)
 int nn_close (int s)
 {
     int rc;
+    struct nn_sock *sock;
 
-    NN_BASIC_CHECKS;
-
-    /*  TODO: nn_sock_term can take a long time to accomplish. It should not be
-        performed under global critical section. */
     nn_glock_lock ();
+    rc = nn_global_hold_socket_locked (&sock, s);
```

Comment:
Because the hold persists -- that is it extends beyond the time when when lock is released.

This paradigm is common in many other code bases:

grab lock.
find object.
bump refcnt (hold the object)
drop the lock

... use the object ...

grab lock
drop reference count
possibly clean up object (we don't really do this quite the same here)
drop lock